src/main/java/com/puppycrawl/tools/checkstyle/MultiThreadChecker.java

```diff
@@ -0,0 +1,226 @@
+
+        try {
+            runAllTasks(tasks);
+        }
+        catch (InterruptedException | ExecutionException ex) {
+            throw new CheckstyleException("Unable to execute checkstyle tasks", ex);
+        }
+        finally {
+            executor.shutdownNow();
+        }
+    }
+
+    @Override
+    protected void finishFileSetChecksProcessing() {
+        super.finishFileSetChecksProcessing();
+        final Set<FileSetCheck> originalFileSetChecks = new HashSet<>(getFileSetChecks());
+        for (FileSetCheck clone : fileSetCheckClones) {
+            if (!originalFileSetChecks.contains(clone)) {
+                clone.finishProcessing();
+            }
+        }
+    }
+
+    @Override
+    protected void destroyFileSetChecks() {
+        super.destroyFileSetChecks();
+        final Set<FileSetCheck> originalFileSetChecks = new HashSet<>(getFileSetChecks());
+        for (FileSetCheck clone : fileSetCheckClones) {
+            if (!originalFileSetChecks.contains(clone)) {
+                clone.destroy();
+            }
+        }
+    }
+
+    /**
+     * Runs all the given tasks.
+     * @param tasks A list of tasks to run.
+     * @throws InterruptedException see {@link MultiThreadChecker#finishTask(Future)}.
+     * @throws CheckstyleException see {@link MultiThreadChecker#finishTask(Future)}.
+     * @throws ExecutionException see {@link MultiThreadChecker#finishTask(Future)}.
+     */
+    private void runAllTasks(List<Callable<ChecksExecutionResult>> tasks)
+            throws InterruptedException, CheckstyleException, ExecutionException {
+
+        final List<Future<ChecksExecutionResult>> futures = tasks.stream()
+                .map(task -> executor.submit(task))
+                .collect(Collectors.toList());
+
+        for (Future<ChecksExecutionResult> future : futures) {
+            finishTask(future);
+        }
+    }
+
+    /**
+     * Finishes the given future.
+     * @param future A future to finish.
+     * @throws InterruptedException might be thrown from {@link Future#get()}.
+     * @throws CheckstyleException might be thrown from a task.
+     * @throws ExecutionException might be thrown from {@link Future#get()}.
+     */
+    private static void finishTask(Future<ChecksExecutionResult> future)
+            throws InterruptedException, CheckstyleException, ExecutionException {
+
+        final ChecksExecutionResult result = future.get();
+        if (result.getThrownException() != null) {
+            throw result.getThrownException();
+        }
+    }
+
+    /**
+     * Creates a callable for processing a particular file.
+     * @param file A file to process.
+     * @return A callable which processes the given file.
+     */
+    private Callable<ChecksExecutionResult> createProcessFileCallable(File file) {
+
+        final String fileName = file.getAbsolutePath();
+
+        return () -> {
+            ChecksExecutionResult result;
+            try {
```

Comment:
Shouldn't this section be a call to something in `AbstractChecker` as it's code looks similar?